File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const set = new Set()
    let l=0, r=0
    let len=0
    while(r < s.length) {
        while(set.has(s[r])) {
            //shrinking window
            set.delete(s[l])
            l++
        }
        set.add(s[r])
        len = Math.max(len, r-l+1)
        r++
    }
    return len
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet();
        int maxlen = 0;
        int l=0, r=0;

        while(r < s.length()) {
            char ch = s.charAt(r);
            if(set.contains(ch)) {
                set.remove(s.charAt(l));
                l++;
                // set = new HashSet();
            } else {
                set.add(ch);
                maxlen = Math.max(maxlen, r-l+1);
                r++;
            }
        }
        return maxlen;
    }
}